File: lua/linked-list/linked-list.lua
LinkedList = {}
LinkedList.__index = LinkedList

function LinkedList:push(value)
	local node = { value = value, prev = self.right.prev, next = self.right }
	self.right.prev.next = node
	self.right.prev = node
end

function LinkedList:unshift(value)
	local node = { value = value, next = self.left.next, prev = self.left }
	self.left.next.prev = node
	self.left.next = node
end

function LinkedList:delete(value)
	local node = self.left
	while node ~= self.right do
		node = node.next
		if node.value == value then
			node.prev.next = node.next
			node.next.prev = node.prev
		end
	end
end

function LinkedList:pop()
	local v = self.right.prev.value
	self.right.prev = self.right.prev.prev
	self.right.prev.next = self.right
	return v
end

function LinkedList:shift()
	local v = self.left.next.value
	self.left.next = self.left.next.next
	self.left.next.prev = self.left
	return v
end

function LinkedList:count()
	local count = 0
	local node = self.left.next
	while node ~= self.right do
		node = node.next
		count = count + 1
	end
	return count
end

return function()
	local left, right = {}, {}
	left.next, right.prev = right, left
	return setmetatable({ left = left, right = right }, LinkedList)
end

File: solutions/php/linked-list/1/LinkedList.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

class LinkedListNode
{
    public function __construct(public int $val, public ?LinkedListNode $next = null, public ?LinkedListNode $previous = null){}
}

class LinkedList
{
    private ?LinkedListNode $front = null;
    private ?LinkedListNode $back = null;

    public function shift(): int
    {
        $node = $this->front;
        $this->front = $this->front->next;
        if($this->front === null) {
            $this->back = null;
        } else {
            $this->front->previous = null;
        }
        return $node->val;
    }

    public function unshift(int $val)
    {
        $node = new LinkedListNode($val, $this->front);
        $this->front = $node;
        if($this->back === null) {
            $this->back = $this->front;
        } else {
            $this->front->next->previous = $this->front;
        }
    }

    public function pop(): int
    {
        $node = $this->back;
        $this->back = $this->back->previous;
        if($this->back === null) {
            $this->front = null;
        } else {
            $this->back->next = null;
        }
        return $node->val;
    }

    public function push(int $val)
    {
        $node = new LinkedListNode($val, null, $this->back);
        $this->back = $node;
        if($this->front === null) {
            $this->front = $this->back;
        } else {
            $this->back->previous->next = $this->back;
        }
    }
}
